framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java

```
@@ -1822,19 +1823,85 @@ protected void commonAssignmentCheck(
+        SimpleAnnotatedTypeScanner<Boolean, Void> checkForMismatchedToStrings =
+                new SimpleAnnotatedTypeScanner<Boolean, Void>() {
+                    /** Maps from a type's toString to its verbose toString */
+                    Map<String, String> map = new HashMap<>();
+
+                    @Override
+                    protected Boolean reduce(Boolean r1, Boolean r2) {
+                        if (r1 == null) {
+                            if (r2 == null) {
+                                return false;
+                            } else {
+                                return r2;
+                            }
+                        } else if (r2 == null) {
+                            return r1;
+                        } else {
+                            return r1 || r2;
+                        }
+                    }
+
+                    @Override
+                    protected Boolean defaultAction(AnnotatedTypeMirror type, Void avoid) {
+                        if (type == null) {
+                            return false;
+                        }
+                        String simple = type.toString();
+                        String verbose = map.get(simple);
+                        if (verbose == null) {
+                            map.put(simple, type.toString(true));
+                            return false;
+                        } else {
+                            return !verbose.equals(type.toString(true));
+                        }
+                    }
+                };
+        Boolean r1 = checkForMismatchedToStrings.visit(atm1);
+        // Call reset to clear the visitor history, but not the map from Strings to types.
+        checkForMismatchedToStrings.reset();
+        Boolean r2 = checkForMismatchedToStrings.visit(atm2);
+
+        // SimpleAnnotatedTypeScanner#scan returns null if it encounters a null AnnotatedTypeMirror.
+        // This shouldn't happen if the atm1 and atm2 are well-formed.
```

If they are ill-formed, I feel it is better for the Checker Framework to abort with an error (so that we get a bug report) than to continue silently.  Is there a reason it's better to ignore the error, such as that it will certainly be detected elsewhere and the CF will halt there?